### DISCUSSION TITLE
Correct way to update uniform buffer? copy from staging or recreate a new uniform buffer?

### DISCUSSION THREAD

**Discussion by shi-yan** [gpuweb/gpuweb#4495] [Q&A]
Hello,

I'm wondering what the correct way of updating a uniform buffer is?

because uniform buffer can't be updated directly, as it doesn't support MAP_WRITE usage. right now, I create another staging buffer, and MAP_WRITE to that staging buffer, then copy the content to the uniform buffer.

is this the right way to do it? it seems to be very verbose. How about recreating a new uniform buffer instead? because when recreating, I can MAP_WRITE to the uniform buffer directly. isn't this faster?

Thanks,

**Comment by greggman** [comment]
AFAIK there is no "correct" way. There are just many different ways. Which way you use is up to you.

Off the top of my head you can

* use `writeBuffer`
* copy to a staging buffer and copy that using `copyBufferToBuffer`
* mark the buffer as both uniform and storage and update it from another shader as a storage buffer
* figure out some creative way to use `copyTextureToBuffer`

> How about recreating a new uniform buffer instead? because when recreating, I can MAP_WRITE to the uniform buffer directly. isn't this faster?

I suspect this is no faster than `writeBuffer` and, in general, slower because of the allocation/deallocation.

**Comment by shaoboyan** [comment]
(More details)
>use writeBuffer
copy to a staging buffer and copy that using copyBufferToBuffer

If the data you transferred is in a small size, I think this is efficient. But when you're trying to upload large (However, in a small chance for a uniform buffer) sized data(e.g. 720p video frame), Pls choose `copyBufferToBuffer`. In current implementation deatils (in chromium), it is much faster.

>How about recreating a new uniform buffer instead? because when recreating, I can MAP_WRITE to the uniform buffer directly. isn't this faster?

This doesn't help from implementation view. When you creating a buffer without MAP_WRITE usage but set `mappedAtCreation` to true. I think browser would help you to create a staging buffer and issue a `copyBufferToBuffer`. And considering the extra allocation/deallocation and some not out-dated data, I don't think this will be faster.

>mark the buffer as both uniform and storage and update it from another shader as a storage buffer

That might be a great idea :)

And if the uniform update is just some "index" info, I'm think DBO helps (https://www.w3.org/TR/webgpu/#programmable-passes).

At last but not least. If the uniform data are small enough, I think community is discussing [push constants](https://github.com/gpuweb/gpuweb/issues/75). Maybe you could follow this issue and consider it as a candidate in future :).

**Reply by shi-yan** [reply]
How can you pass a 720p video frame to a uniform buffer? I thought uniform buffer had a size limit?

**Reply by kainino0x** [reply]
All of this advice applies to any upload (storage buffers, textures, vertex buffers, etc), not just uniform buffer uploads.

Uniform buffers do have a low default limit (https://gpuweb.github.io/gpuweb/#dom-supported-limits-maxuniformbufferbindingsize) but implementations are allowed to support higher limits. Some devices support effectively unlimited uniform buffer sizes.

**Comment by shubham763** [comment]
Both approaches have their advantages and trade-offs, and the choice between them depends on various factors such as performance considerations, memory usage, and the frequency of updates.

So, if you are frequently updating small amounts of data, using a staging buffer may be more efficient despite the additional steps involved. However, if your updates are infrequent or involve large amounts of data, recreating the uniform buffer may be a more suitable approach. It's essential to profile and benchmark your application to determine which approach provides the best performance for your specific use case. Additionally, consider using synchronization techniques such as fences or command dependencies to optimize CPU-GPU synchronization and minimize overhead.
